### PR TITLE
Properly compare velocypack objects in Agency operations

### PR DESCRIPTION
**Pull request by graetzer**
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1587/

### CODE REVIEW

File: 3rdParty/velocypack/src/Dumper.cpp [hunk 1]
```diff
@@ -376,7 +376,7 @@ void Dumper::dumpValue(Slice const* slice, Slice const* base) {
     }
 
     case ValueType::Object: {
-      ObjectIterator it(*slice);
+      ObjectIterator it(*slice, true);
```
Comment (neunhoef) [inline 3rdParty/velocypack/src/Dumper.cpp]: This could have huge consequences in lots of places. I do not know if we can do this...

Comment (neunhoef) [reply 3rdParty/velocypack/src/Dumper.cpp]: Also, if we do this here, we should also do it in the VelocyPack repository...

Comment (graetzer) [reply 3rdParty/velocypack/src/Dumper.cpp]: reverted that change in 3.3, added it to the changelog >= 3.4

File: arangod/Agency/Node.cpp [hunk 2]
```diff
@@ -581,8 +582,13 @@ bool Node::handle<REPLACE>(VPackSlice const& slice) {
   Builder tmp;
   { VPackArrayBuilder t(&tmp);
     if (this->slice().isArray()) {
+      VPackSlice valToRepl = slice.get("val");
       for (auto const& old : VPackArrayIterator(this->slice())) {
-        tmp.add(old == slice.get("val") ? slice.get("new") : old);
+        if (VelocyPackHelper::compare(old, valToRepl, /*useUTF8*/true) == 0) {
```
Comment (jsteemann) [inline arangod/Agency/Node.cpp]: We are checking for equality only here, not for < or >. Can't we set `useUtf8` to false then?

File: arangod/Agency/Node.cpp [hunk 3]
```diff
@@ -539,8 +539,9 @@ template <> bool Node::handle<ERASE>(VPackSlice const& slice) {
 
     if (this->slice().isArray()) {
       if (haveVal) {
+        VPackSlice valToErase = slice.get("val");
         for (auto const& old : VPackArrayIterator(this->slice())) {
-          if (old != slice.get("val")) {
+          if (VelocyPackHelper::compare(old, valToErase, /*useUTF8*/true) != 0) {
```
Comment (jsteemann) [inline arangod/Agency/Node.cpp]: We are checking for equality only here, not for < or >. Can't we set `useUtf8` to false then?

File: 3rdParty/velocypack/src/Dumper.cpp (hunk 1 above)
Comment (jsteemann) [inline 3rdParty/velocypack/src/Dumper.cpp]: I would prefer making this a flag in `velocypack::Options`, which can set to "dump in any order" by default in ArangoDB.

File: arangod/Agency/Node.cpp (hunk 2 above)
Comment (graetzer) [reply arangod/Agency/Node.cpp]: I don't know technically the docs say just equality, so unicode equality would have to be accounted for ?

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/1610/

**Pull request merged**